# 04 – Details

Esplorazione e data cleaning del dataset `details.csv`.

**Colonne:**
| Colonna | Descrizione |
|---|---|
| `mal_id` | ID univoco dell'anime su MyAnimeList |
| `title` | Titolo dell'anime |
| `title_japanese` | Titolo in giapponese |
| `url` | URL della pagina MAL dell'anime |
| `image_url` | URL dell'immagine di copertina |
| `type` | Tipo (TV, Movie, etc) |
| `status` | Stato di trasmissione |
| `score` | Punteggio medio |
| `scored_by` | Numero di utenti che hanno votato |
| `start_date` | Data di inizio trasmissione |
| `end_date` | Data di fine trasmissione |
| `synopsis` | Sinossi testuale |
| `rank` | Classifica globale |
| `popularity` | Popolarità |
| `members` | Numero di utenti che hanno l'anime nella loro lista |
| `favorites` | Numero di utenti che hanno messo l'anime tra i preferiti |
| `genres` | Lista di generi |
| `studios` | Lista di studi di produzione |
| `themes` | Lista di temi narrativi |
| `demographics` | Lista di target demografici |
| `source` | Materiale di origine (Manga, Novel, Original, etc) |
| `rating` | Classificazione per età |
| `episodes` | Numero di episodi |
| `season` | Stagione di uscita |
| `year` | Anno di uscita |
| `producers` | Lista di produttori |
| `explicit_genres` | Lista di generi espliciti |
| `licensors` | Lista di licenziatari |
| `streaming` | Lista di piattaforme di streaming |

## 1. Import e caricamento dati
Importiamo le librerie necessarie e carichiamo il file csv. Facciamo una esplorazione generica per capire la struttura e le caratteristiche del dataset.

In [1]:
import pandas as pd
import numpy as np
from dataset_analyzer import analyze

df_details = pd.read_csv('../datasets/details.csv')
print(f'Shape: {df_details.shape}')
df_details.info()
df_details.head()

Shape: (28955, 29)
<class 'pandas.DataFrame'>
RangeIndex: 28955 entries, 0 to 28954
Data columns (total 29 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   mal_id           28955 non-null  int64  
 1   title            28955 non-null  str    
 2   title_japanese   28836 non-null  str    
 3   url              28955 non-null  str    
 4   image_url        28955 non-null  str    
 5   type             28888 non-null  str    
 6   status           28955 non-null  str    
 7   score            18882 non-null  float64
 8   scored_by        18882 non-null  float64
 9   start_date       28104 non-null  str    
 10  end_date         11267 non-null  str    
 11  synopsis         23828 non-null  str    
 12  rank             21997 non-null  float64
 13  popularity       28955 non-null  int64  
 14  members          28955 non-null  int64  
 15  favorites        28955 non-null  int64  
 16  genres           28955 non-null  str    
 17  stud

,mal_id,title,title_japanese,url,image_url,type,status,score,scored_by,start_date,...,demographics,source,rating,episodes,season,year,producers,explicit_genres,licensors,streaming
0,59356,-Socket-,-socket-,https://myanimelist.net/anime/59356/-Socket-,https://cdn.myanimelist.net/images/anime/1043/...,Movie,Finished Airing,NaN,NaN,2010-01-01T00:00:00+00:00,...,[],Original,G - All Ages,1.0,NaN,NaN,['Nagoya Zokei University'],[],[],[]
1,56036,......,......,https://myanimelist.net/anime/56036/-,https://cdn.myanimelist.net/images/anime/1057/...,Music,Finished Airing,6.53,503.0,2023-06-11T00:00:00+00:00,...,[],Original,PG-13 - Teens 13 or older,1.0,NaN,NaN,[],[],[],[]
2,2928,.hack//G.U. Returner,.HACK//G.U. RETURNER,https://myanimelist.net/anime/2928/hack__GU_Re...,https://cdn.myanimelist.net/images/anime/1798/...,OVA,Finished Airing,6.65,9745.0,2007-01-18T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,"['Bandai Visual', 'CyberConnect2']",[],[],[]
3,3269,.hack//G.U. Trilogy,.hack//G.U. Trilogy,https://myanimelist.net/anime/3269/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/1566/...,Movie,Finished Airing,7.06,15373.0,2007-12-22T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,['Bandai Visual'],[],"['Funimation', 'Bandai Entertainment']",[]
4,4469,.hack//G.U. Trilogy: Parody Mode,.hack//G.U. Trilogy,https://myanimelist.net/anime/4469/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/10/86...,Special,Finished Airing,6.35,4317.0,2008-03-25T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,['Bandai Visual'],[],[],[]


Il dataset contiene **28.955 righe** e **29 colonne**. Alcune colonne presentano valori nulli. Alcuni dtype richiedono conversione: le date sono in formato stringa, e alcune colonne intere (`scored_by`, `rank`, `episodes`, `year`) sono memorizzate come `float64`. Quiesti veranno gestiti singolarmente nella sezione 2. 

## 1.1 Rimozione duplicati esatti

Prima dell'analisi per colonna, rimuoviamo le righe con valori identici in **tutte** le colonne, mantenendo solo la prima occorrenza.

In [2]:
n_originale = len(df_details)

mask_dup = df_details.duplicated(keep=False)       # tutte le occorrenze coinvolte
n_righe_coinvolte = mask_dup.sum()                  # righe totali che hanno almeno un duplicato
n_gruppi = df_details[mask_dup].duplicated(keep='first').sum()  # occorrenze extra (da rimuovere)
n_tenute = n_righe_coinvolte - n_gruppi             # prime occorrenze mantenute

print(f'Righe totali coinvolte in duplicazioni : {n_righe_coinvolte:,}')
print(f'  → prime occorrenze mantenute         : {n_tenute:,}')
print(f'  → occorrenze extra rimosse           : {n_gruppi:,}')
print()

df_details.drop_duplicates(keep='first', inplace=True)
print(f'Righe prima della rimozione : {n_originale:,}')
print(f'Righe dopo la rimozione     : {len(df_details):,}')

Righe totali coinvolte in duplicazioni : 0
  → prime occorrenze mantenute         : 0
  → occorrenze extra rimosse           : 0



Righe prima della rimozione : 28,955
Righe dopo la rimozione     : 28,955


## 2. Analisi colonna per colonna

### 2.1 `mal_id`

È la **chiave primaria** del dataset. Deve essere intera, non nulla e univoca. Utilizziamo `analyze` per verificare queste proprietà.

In [3]:
analyze(df_details['mal_id'])


  Nome serie                     mal_id
  dtype                          int64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Zeri                           0  (0.00%)
  Positivi                       28,955   (100.00%)
  Negativi                       0   (0.00%)
  Valori unici                   28,955  (100.00%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1
  Max                            62,590
  Range                          62,589
  Media                          33,977.52
  Mediana                        37,969.00
  Moda/e                         1, 5, 6
  Dev. standard    

  I 10 più frequenti:
                Valore    Conteggio        %
  ────────────────────   ──────────   ──────
                59,356            1     0.0%
                56,036            1     0.0%
                 2,928            1     0.0%
                 3,269            1     0.0%
                 4,469            1     0.0%
                   454            1     0.0%
                 1,143            1     0.0%
                   299            1     0.0%
                 9,332            1     0.0%
                10,390            1     0.0%

  I 10 meno frequenti:
                Valore    Conteggio        %
  ────────────────────   ──────────   ──────
                56,606            1     0.0%
                34,895            1     0.0%
                36,305            1     0.0%
                31,245            1     0.0%
                59,421            1     0.0%
                58,303            1     0.0%
                52,512            1     0.0%
         

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un ID univoco.
- **Nessun duplicato**: ogni riga corrisponde a un anime distinto.
- **Dtype corretto**: `int64`, nessuna conversione necessaria.

Nessuna pulizia necessaria.

### 2.2 `title`

Colonna testuale con il titolo dell'anime. Verifichiamo la presenza di valori nulli e duplicati.

In [4]:
analyze(df_details['title'])


  Nome serie                     title
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   28,953  (99.99%)
  Valori duplicati               2

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  1  caratteri
  Lunghezza max                  186  caratteri
  Lunghezza media                24.70  caratteri
  Lunghezza mediana              21.0000  caratteri
  Dev. standard lunghezza        15.07
  IQR lunghezza                  18.0000

Valori di lunghezza estrema
───────────────────

  Contiene cifre                 3,715  (12.8%)
  Contiene maiuscole             28,915  (99.9%)
  Contiene minuscole             28,861  (99.7%)
  Contiene spazi                 26,023  (89.9%)
  Contiene caratteri speciali    12,572  (43.4%)
  Contiene URL                   1  (0.0%)
  Contiene @                     53  (0.2%)
  Solo numerico                  14
  Solo alfabetico                2,302
  Solo alfanumerico              2,332

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────



  Caratteri totali               715,147
  Caratteri unici                165
  Caratteri medi per valore      24.70

Analisi parole / token
────────────────────────────────────────────────────────────────────────────────



  Parole medie per valore        4.08
  Max parole                     30
  Min parole                     1

  Le 10 parole più comuni:
  Parola                  Conteggio
  ────────────────────   ──────────
  no                          6,572
  to                          1,207
  Season                      1,051
  ni                            817
  wa                            777
  -                             727
  The                           713
  Movie                         641
  wo                            567
  Xiao                          558



**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un titolo.
- **Quasi univoco**: 28.953 valori distinti su 28.955. Ci solo 2 titoli duplicati che verifichiamo successivamente. 
- **Lunghezze variabili**: da 1 carattere (titoli come `2`, `9`, singole lettere) a 186 caratteri (lunghi titoli isekai). Vacciamo delle verifiche per approfondire. 
- **14 titoli solo numerici** (es. `2`, `9`). Indaghiamo in seguito.
- **1 titolo contiene un URL** (~0.0%). Ingaghiamo in seguito.
- **43.4% contiene caratteri speciali**: normale, i titoli includono punteggiatura, simboli e trattini.

In [5]:
# Titoli duplicati 
dup_mask = df_details.duplicated(subset=['title'], keep=False)
print(f'Titoli duplicati (righe coinvolte): {dup_mask.sum()}')
print()
print(df_details[dup_mask][['mal_id', 'title', 'type', 'status', 'year']].to_string(index=False))

# Titoli solo numerici 
print()
print('─' * 60)
num_mask = df_details['title'].str.fullmatch(r'\d+')
print(f'Titoli solo numerici: {num_mask.sum()}')
print()
print(df_details[num_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

# Titolo che contiene un URL 
print()
print('─' * 60)
url_mask = df_details['title'].str.contains(r'https?://', regex=True, na=False)
print(f'Titoli con URL: {url_mask.sum()}')
print()
print(df_details[url_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

# Lunghezze estreme 
print()
print('─' * 60)
print('Titoli più corti (len ≤ 2):')
short_mask = df_details['title'].str.len() <= 2
print(df_details[short_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))
print()
print('Titoli più lunghi (len ≥ 100):')
long_mask = df_details['title'].str.len() >= 100
print(df_details[long_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

Titoli duplicati (righe coinvolte): 4

 mal_id                                    title  type          status  year
  59994                                Moonlight Music Finished Airing   NaN
  60524                                Moonlight Music Finished Airing   NaN
  53367 Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian   ONA Finished Airing   NaN
  55740 Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian   ONA Finished Airing   NaN

────────────────────────────────────────────────────────────
Titoli solo numerici: 14

 mal_id  title  type   year
  29978    001   ONA    NaN
  60145   1135 Music    NaN
  37065    123 Music    NaN
  55204   1985 Music    NaN
  30234   1989 Music    NaN
  38756      2 Music    NaN
  54066   2005 Music    NaN
  32813   2010 Music    NaN
  53789     32 Music    NaN
  53771   3341 Music    NaN
  60486    520 Movie    NaN
  13971 663114 Movie    NaN
  41457     86    TV 2021.0
  36089      9 Music    NaN

────────────────────────────────────────────────────────────
Titoli

**Risultati dell'indagine:**

- **Titoli duplicati (2 coppie, 4 righe):** `Moonlight` e `Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian` compaiono ciascuno due volte, ma con `mal_id` distinti. Sono anime separati su MAL che condividono il titolo. **Nessuna rimozione necessaria.**

- **Titoli solo numerici (14 righe):** tutti legittimi. La maggior parte sono video musicali (`Music`) o cortometraggi in cui titoli brevi o numerici sono comuni. **Nessuna azione necessaria.**

- **Titolo contenente un URL (1 riga):** `Himitsukessha Taka no Tsume The Movie 3: http://takanotsume.jp wa Eien ni` (`mal_id` 8357). L'URL è parte integrante del titolo ufficiale del film, caratteristica della serie *Taka no Tsume* nota per titoli volutamente assurdi. **Nessuna azione necessaria.**

- **Lunghezze estreme:** i titoli molto corti (1–2 caratteri) sono anime reali con titoli brevi; i titoli molto lunghi (fino a 186 caratteri) sono tipici degli produzioni moderni. Nessun valore anomalo.

**Conclusione:** la colonna `title` non richiede alcuna pulizia. Tutti i casi apparentemente anomali sono valori legittimi.

### 2.3 `title_japanese`

Colonna testuale con il titolo in giapponese. I null sono ammessi: anime di produzione non giapponese o senza titolo kanji registrato su MAL non avranno questo campo.

In [6]:
analyze(df_details['title_japanese'])


  Nome serie                     title_japanese
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,836  (99.59%)
  Null / NaN                     119  (0.41%)
  Stringhe vuote                 0
  Valori unici                   27,772  (95.91%)
  Valori duplicati               1,064

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  1  caratteri
  Lunghezza max                  85  caratteri
  Lunghezza media                12.24  caratteri
  Lunghezza mediana              10.0000  caratteri
  Dev. standard lunghezza        8.69
  IQR lunghezza                  10.0000

Valori di lunghezza estrema
───────

  Contiene cifre                 2,618  (9.1%)
  Contiene maiuscole             6,097  (21.1%)
  Contiene minuscole             3,300  (11.4%)
  Contiene spazi                 10,743  (37.3%)
  Contiene caratteri speciali    27,304  (94.7%)
  Contiene URL                   0  (0.0%)
  Contiene @                     10  (0.0%)
  Solo numerico                  9
  Solo alfabetico                598
  Solo alfanumerico              621

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────

  Caratteri totali               352,903
  Caratteri unici                4,559
  Caratteri medi per valore      12.24



Analisi parole / token
────────────────────────────────────────────────────────────────────────────────

  Parole medie per valore        1.65
  Max parole                     14
  Min parole                     1

  Le 10 parole più comuni:
  Parola                  Conteggio
  ────────────────────   ──────────
  THE                           418
  ANIMATION                     237
  劇場版                           231
  The                           186
  第二季                           160
  of                            140
  the                           136
  2                             131
  映画                            120
  -                              92



**Osservazioni:**
- **119 valori nulli** (~0.4%): strutturali, anime di produzione non giapponese o senza titolo kanji registrato su MAL.

Nessuna pulizia necessaria.

In [7]:
# Nessuna operazione richiesta: i null sono strutturali
print(f'Null in title_japanese : {df_details["title_japanese"].isna().sum()} ({df_details["title_japanese"].isna().mean()*100:.1f}%)')

Null in title_japanese : 119 (0.4%)


**Verifica caratteri Est-Asiatici in `title_japanese`**

I valori non-null dovrebbero contenere caratteri giapponesi, cinesi o coreani. Verifichiamo che non ci siano valori non-null privi di tali caratteri, che potrebbero indicare errori di inserimento.

In [8]:
non_null_tj = df_details['title_japanese'].dropna()
has_ea = non_null_tj.str.contains(
    r'[\u1100-\u11ff\u3000-\u9fff\uac00-\ud7af\uff00-\uffef]', regex=True
)
non_ea_tj = non_null_tj[~has_ea]
print(f'title_japanese non-null senza caratteri Est-Asiatici: {len(non_ea_tj)}')
print(non_ea_tj.value_counts().head(10).to_string())

title_japanese non-null senza caratteri Est-Asiatici: 2135
title_japanese
HiGH&LOW g-sword            3
Wind -a breath of heart-    3
.hack//G.U. Trilogy         2
7SEEDS                      2
AIKa ZERO                   2
AMNESIA                     2
Awakening                   2
BAD MOOD                    2
BT21 ORIGINAL STORY         2
CØDE:BREAKER                2


**Osservazioni:**
- **2.135 valori** non-null privi di caratteri Est-Asiatici (~7.4% dei non-null): si tratta di anime il cui titolo ufficiale giapponese è già in caratteri latini o ASCII (es. `AMNESIA`, `7SEEDS`, `.hack//G.U. Trilogy`). È un comportamento atteso su MAL. **Nessuna pulizia necessaria.**

### 2.4 `url`

URL della pagina MAL dell'anime. Essendo derivato direttamente dal `mal_id`, ci aspettiamo assenza di null e unicità. Verifichiamo anche che il format sia corretto.

In [9]:
analyze(df_details['url'])

pattern_url = r'^https://myanimelist\.net/anime/\d+/\S+$'
non_conformi_url = df_details[~df_details['url'].str.match(pattern_url, na=False)]
print(f'URL non conformi al formato atteso: {len(non_conformi_url)}')

# Consistenza mal_id <-> url
id_from_url = df_details['url'].str.extract(r'/anime/(\d+)/')[0].astype(int)
mismatch_url = id_from_url != df_details['mal_id']
print(f'mal_id non coerente con url        : {mismatch_url.sum()}')


  Nome serie                     url
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   28,955  (100.00%)
  Valori duplicati               0

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  35  caratteri
  Lunghezza max                  216  caratteri
  Lunghezza media                60.19  caratteri
  Lunghezza mediana              57.0000  caratteri
  Dev. standard lunghezza        14.92
  IQR lunghezza                  17.0000

Valori di lunghezza estrema
───────────────────

  Contiene cifre                 28,955  (100.0%)
  Contiene maiuscole             28,915  (99.9%)
  Contiene minuscole             28,955  (100.0%)
  Contiene spazi                 0  (0.0%)
  Contiene caratteri speciali    28,955  (100.0%)
  Contiene URL                   28,955  (100.0%)
  Contiene @                     0  (0.0%)
  Solo numerico                  0
  Solo alfabetico                0
  Solo alfanumerico              0

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────

  Caratteri totali               1,742,829
  Caratteri unici                145
  Caratteri medi per valore      60.19

URL non conformi al formato atteso: 0
mal_id non coerente con url        : 0


**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un URL valido.
- **Nessun duplicato**: coerente con `mal_id` come chiave primaria.


**Verifica formato e consistenza:**
- Tutti i 28.955 URL seguono il pattern `https://myanimelist.net/anime/{id}/...`. Nessuna anomalia.
- L'ID estratto dall'URL corrisponde esattamente a `mal_id` per ogni riga. Nessuna inconsistenza.

Nessuna pulizia necessaria.


### 2.5 `image_url`

URL dell'immagine di copertina. Verifichiamo anche il format del URL. 

In [10]:
DEFAULT_IMG = 'https://cdn.myanimelist.net/img/sp/icon/apple-touch-icon-256.png'
placeholder = (df_details['image_url'] == DEFAULT_IMG).sum()
print(f'Immagini placeholder : {placeholder:,} ({placeholder/len(df_details)*100:.1f}%)')
print(f'Immagini reali       : {len(df_details)-placeholder:,} ({(len(df_details)-placeholder)/len(df_details)*100:.1f}%)')
print()
analyze(df_details['image_url'])

pattern_img = r'^https://cdn\.myanimelist\.net/(images/anime/\d+/\d+\.(jpg|webp|png)|img/sp/icon/apple-touch-icon-256\.png)$'
non_conformi_img = df_details[~df_details['image_url'].str.match(pattern_img, na=False)]
print(f'image_url non conformi al formato atteso: {len(non_conformi_img)}')

Immagini placeholder : 198 (0.7%)
Immagini reali       : 28,757 (99.3%)


  Nome serie                     image_url
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   28,758  (99.32%)
  Valori duplicati               197

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  50  caratteri
  Lunghezza max                  64  caratteri
  Lunghezza media                54.69  caratteri
  Lunghezza mediana              56.0000  caratteri
  Dev. standard lunghezza        1.93
  IQR lungh

  I 10 più frequenti:
  Valore                                                              Conteggio        %
  ────────────────────────────────────────────────────────────────   ──────────   ──────
  https://cdn.myanimelist.net/img/sp/icon/apple-touch-icon-256.png          198     0.7%
  https://cdn.myanimelist.net/images/anime/1043/144360.jpg                    1     0.0%
  https://cdn.myanimelist.net/images/anime/1057/137264.jpg                    1     0.0%
  https://cdn.myanimelist.net/images/anime/1798/115989.jpg                    1     0.0%
  https://cdn.myanimelist.net/images/anime/1566/151023.jpg                    1     0.0%
  https://cdn.myanimelist.net/images/anime/10/8661.jpg                        1     0.0%
  https://cdn.myanimelist.net/images/anime/2/23085.jpg                        1     0.0%
  https://cdn.myanimelist.net/images/anime/1844/131542.jpg                    1     0.0%
  https://cdn.myanimelist.net/images/anime/7/23079.jpg                        1     0.0%

  Contiene cifre                 28,955  (100.0%)
  Contiene maiuscole             0  (0.0%)
  Contiene minuscole             28,955  (100.0%)
  Contiene spazi                 0  (0.0%)
  Contiene caratteri speciali    28,955  (100.0%)
  Contiene URL                   28,955  (100.0%)
  Contiene @                     0  (0.0%)
  Solo numerico                  0
  Solo alfabetico                0
  Solo alfanumerico              0

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────



  Caratteri totali               1,583,516
  Caratteri unici                31
  Caratteri medi per valore      54.69

image_url non conformi al formato atteso: 0


**Osservazioni:**
- Tutti i 28.955 URL seguono il pattern corretto: immagini reali `cdn.myanimelist.net/images/anime/...` e placeholder `apple-touch-icon-256.png`. Nessuna anomalia.
- Si nota che 198 URL corrispondono all'immagine generica di MAL, segnale di assenza di copertina reale che sarebbe un placeholder. Aggiungiamo una colonna booleana `has_image` per distinguere tra gli anime con e senza imagine. 

In [11]:
DEFAULT_IMG = 'https://cdn.myanimelist.net/img/sp/icon/apple-touch-icon-256.png'
df_details['has_image'] = df_details['image_url'] != DEFAULT_IMG
print(f'has_image → True (reale): {df_details["has_image"].sum():,}  |  False (placeholder): {(~df_details["has_image"]).sum():,}')

has_image → True (reale): 28,757  |  False (placeholder): 198


### 2.6 `type`

Colonna categorica con il tipo di anime (TV, Movie, OVA, ONA, Special, Music, CM, PV, TV Special). I null sono attesi per gli anime `Not yet aired` il cui tipo non è ancora registrato su MAL.

In [12]:
analyze(df_details['type'])


  Nome serie                     type
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,888  (99.77%)
  Null / NaN                     67  (0.23%)
  Stringhe vuote                 0
  Valori unici                   9  (0.03%)
  Valori duplicati               28,879

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  2  caratteri
  Lunghezza max                  10  caratteri
  Lunghezza media                3.73  caratteri
  Lunghezza mediana              3.0000  caratteri
  Dev. standard lunghezza        1.81
  IQR lunghezza                  3.0000

Valori di lunghezza estrema
──────────────────────────

  Contiene cifre                 0  (0.0%)
  Contiene maiuscole             28,888  (100.0%)
  Contiene minuscole             11,436  (39.6%)
  Contiene spazi                 767  (2.7%)
  Contiene caratteri speciali    0  (0.0%)
  Contiene URL                   0  (0.0%)
  Contiene @                     0  (0.0%)
  Solo numerico                  0
  Solo alfabetico                28,121
  Solo alfanumerico              28,121

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────

  Caratteri totali               107,709
  Caratteri unici                20
  Caratteri medi per valore      3.73



**Osservazioni:**
- **67 valori nulli**: strutturali, anime con stato `Not yet aired` per cui il tipo non è ancora registrato su MAL.
- **9 valori distinti**: tutti coerenti con la tassonomia MAL (TV, Movie, OVA, ONA, Special, Music, CM, PV, TV Special).

Nessuna pulizia necessaria.

### 2.7 `status`

Colonna categorica con lo stato di trasmissione. I valori attesi sono `Finished Airing`, `Currently Airing` e `Not yet aired`.

In [13]:
analyze(df_details['status'])


  Nome serie                     status
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   3  (0.01%)
  Valori duplicati               28,952

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────



  Lunghezza min                  13  caratteri
  Lunghezza max                  16  caratteri
  Lunghezza media                14.97  caratteri
  Lunghezza mediana              15.0000  caratteri
  Dev. standard lunghezza        0.29
  IQR lunghezza                  0.0000

Valori di lunghezza estrema
────────────────────────────────────────────────────────────────────────────────

  Le 10 stringhe più lunghe:
    #1   len=16  Currently Airing
    #2   len=16  Currently Airing
    #3   len=16  Currently Airing
    #4   len=16  Currently Airing
    #5   len=16  Currently Airing
    #6   len=16  Currently Airing
    #7   len=16  Currently Airing
    #8   len=16  Currently Airing
    #9   len=16  Currently Airing
    #10  len=16  Currently Airing

  Le 10 stringhe più corte:
    #1   len=13  Not yet aired
    #2   len=13  Not yet aired
    #3   len=13  Not yet aired
    #4   len=13  Not yet aired
    #5   len=13  Not yet aired
    #6   len=13  Not yet aired
    #7   len=13  Not yet aired


  I 10 più frequenti:
  Valore              Conteggio        %
  ────────────────   ──────────   ──────
  Finished Airing        28,097    97.0%
  Not yet aired             544     1.9%
  Currently Airing          314     1.1%

  I 10 meno frequenti (più rari):
  Valore              Conteggio        %
  ────────────────   ──────────   ──────
  Currently Airing          314     1.1%
  Not yet aired             544     1.9%
  Finished Airing        28,097    97.0%

Analisi pattern / contenuto
────────────────────────────────────────────────────────────────────────────────



  Contiene cifre                 0  (0.0%)
  Contiene maiuscole             28,955  (100.0%)
  Contiene minuscole             28,955  (100.0%)
  Contiene spazi                 28,955  (100.0%)
  Contiene caratteri speciali    0  (0.0%)
  Contiene URL                   0  (0.0%)
  Contiene @                     0  (0.0%)
  Solo numerico                  0
  Solo alfabetico                0
  Solo alfanumerico              0

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────

  Caratteri totali               433,551
  Caratteri unici                19
  Caratteri medi per valore      14.97



Analisi parole / token
────────────────────────────────────────────────────────────────────────────────



  Parole medie per valore        2.02
  Max parole                     3
  Min parole                     2



  Le 10 parole più comuni:
  Parola                  Conteggio
  ────────────────────   ──────────
  Airing                     28,411
  Finished                   28,097
  Not                           544
  yet                           544
  aired                         544
  Currently                     314



**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno uno stato definito.
- **3 valori distinti**: `Finished Airing`, `Currently Airing`, `Not yet aired`, coerenti con la tassonomia MAL.

Nessuna pulizia necessaria.

In [14]:
# Nessuna operazione richiesta
print(f'Null in status : {df_details["status"].isna().sum()}')
print(df_details['status'].value_counts())

Null in status : 0
status
Finished Airing     28097
Not yet aired         544
Currently Airing      314
Name: count, dtype: int64


### 2.8 `score` e `scored_by`

Le due colonne sono co-nulle: quando un anime non ha voti sufficienti su MAL, entrambe sono `NaN`.

In [15]:
analyze(df_details['score'])


  Nome serie                     score
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               18,882  (65.21%)
  Null / NaN                     10,073  (34.79%)
  Zeri                           0  (0.00%)
  Positivi                       18,882   (65.21%)
  Negativi                       0   (0.00%)
  Valori unici                   561  (1.94%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1.8900
  Max                            9.2900
  Range                          7.4000
  Media                          6.3905
  Mediana                        6.3600
  Moda/e                         6.2900
  Dev. standard      

In [16]:
analyze(df_details['scored_by'])


  Nome serie                     scored_by
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               18,882  (65.21%)
  Null / NaN                     10,073  (34.79%)
  Zeri                           0  (0.00%)
  Positivi                       18,882   (65.21%)
  Negativi                       0   (0.00%)
  Valori unici                   9,245  (31.93%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            101.0000
  Max                            2,979,733
  Range                          2,979,632
  Media                          29,963.08
  Mediana                        1,528.00
  Moda/e                         120.0000

**Osservazioni:**
- **10.073 valori nulli** (~34.8%) in entrambe le colonne: strutturali, anime senza voti sufficienti per avere uno score su MAL. Co-nullità confermata.
- **`score`**: dtype `float64` corretto (valori decimali). Nessuna conversione necessaria.
- **`scored_by`**: dtype `float64`. Lo convertiamo a `Int64`.

In [17]:
# Verifica co-nullità
print(f'Null in score     : {df_details["score"].isna().sum():,}')
print(f'Null in scored_by : {df_details["scored_by"].isna().sum():,}')
print(f'Co-nulli (entrambi NaN) : {(df_details["score"].isna() & df_details["scored_by"].isna()).sum():,}')
print()

df_details['scored_by'] = df_details['scored_by'].astype('Int64')
print(f'scored_by dtype : {df_details["scored_by"].dtype}')

Null in score     : 10,073
Null in scored_by : 10,073
Co-nulli (entrambi NaN) : 10,073

scored_by dtype : Int64


### 2.9 `start_date`

Data di inizio trasmissione, memorizzata come stringa. Usiamo `analyze` per capire il formato e valutare la conversione in un tipo datetime. I null sono attesi per gli anime `Not yet aired` senza data confermata.

In [18]:
analyze(df_details['start_date'])


  Nome serie                     start_date
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,104  (97.06%)
  Null / NaN                     851  (2.94%)
  Stringhe vuote                 0
  Valori unici                   9,367  (32.35%)
  Valori duplicati               18,737



Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  25  caratteri
  Lunghezza max                  25  caratteri
  Lunghezza media                25.00  caratteri
  Lunghezza mediana              25.0000  caratteri
  Dev. standard lunghezza        0.00
  IQR lunghezza                  0.0000

Valori di lunghezza estrema
────────────────────────────────────────────────────────────────────────────────

  Le 10 stringhe più lunghe:
    #1   len=25  2010-01-01T00:00:00+00:00
    #2   len=25  2023-06-11T00:00:00+00:00
    #3   len=25  2007-01-18T00:00:00+00:00
    #4   len=25  2007-12-22T00:00:00+00:00
    #5   len=25  2008-03-25T00:00:00+00:00
    #6   len=25  2003-11-16T00:00:00+00:00
    #7   len=25  2003-03-28T00:00:00+00:00
    #8   len=25  2002-06-20T00:00:00+00:00
    #9   len=25  2011-01-28T00:00:00+00:00
    #10  len=25  2011-01-28T00:00:00+00:00

  Le 10 stringhe più corte:
    #1   len=25  

  Contiene cifre                 28,104  (100.0%)


  Contiene maiuscole             28,104  (100.0%)
  Contiene minuscole             0  (0.0%)
  Contiene spazi                 0  (0.0%)
  Contiene caratteri speciali    28,104  (100.0%)
  Contiene URL                   0  (0.0%)
  Contiene @                     0  (0.0%)
  Solo numerico                  0
  Solo alfabetico                0
  Solo alfanumerico              0

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────

  Caratteri totali               702,600
  Caratteri unici                14
  Caratteri medi per valore      25.00



**Osservazioni:**
- **851 valori nulli** (~2.9%): strutturali, anime `Not yet aired` senza data confermata.
- **Lunghezza costante di 25 caratteri** e tutti gli esempi mostrano il pattern `YYYY-MM-DDTHH:MM:SS+00:00`: che corrisponde al formato ISO 8601 con offset `+00:00` che rapresenta il timezone.
- **Conversione necessaria**: da stringa a `datetime64[us, UTC]` per abilitare operazioni temporali (confronti, differenze, etc). Usiamo UTC perché è l'unica scelta coerente con l'offset `+00:00` già presente nei dati.

In [19]:
df_details['start_date'] = pd.to_datetime(df_details['start_date'], errors='coerce', utc=True)
print(f'start_date dtype : {df_details["start_date"].dtype}')
print(f'Null residui     : {df_details["start_date"].isna().sum()}')
print(f'Range            : {df_details["start_date"].min()} → {df_details["start_date"].max()}')

start_date dtype : datetime64[us, UTC]
Null residui     : 851


Range            : 1917-01-01 00:00:00+00:00 → 2027-01-01 00:00:00+00:00


### 2.10 `end_date`

Data di fine trasmissione, stessa struttura di `start_date`. I null sono più numerosi: gli anime `Currently Airing` e `Not yet aired` non hanno ancora una data di fine.

In [20]:
analyze(df_details['end_date'])


  Nome serie                     end_date
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               11,267  (38.91%)
  Null / NaN                     17,688  (61.09%)
  Stringhe vuote                 0
  Valori unici                   5,607  (19.36%)
  Valori duplicati               5,660

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  25  caratteri
  Lunghezza max                  25  caratteri
  Lunghezza media                25.00  caratteri
  Lunghezza mediana              25.0000  caratteri
  Dev. standard lunghezza        0.00
  IQR lunghezza                  0.0000

Valori di lunghezza estrema
──────────

**Osservazioni:**
- **17.688 valori nulli** (~61.1%): strutturali, anime ancora in corso o non ancora iniziati.
- Il format è lo stesso di `start_date`. Facciamo la conversione a `datetime64[us, UTC]`.

In [21]:
df_details['end_date'] = pd.to_datetime(df_details['end_date'], errors='coerce', utc=True)
print(f'end_date dtype : {df_details["end_date"].dtype}')
print(f'Null residui   : {df_details["end_date"].isna().sum()}')
print(f'Range          : {df_details["end_date"].min()} → {df_details["end_date"].max()}')

end_date dtype : datetime64[us, UTC]
Null residui   : 17688
Range          : 1962-02-25 00:00:00+00:00 → 2026-04-10 00:00:00+00:00


**Verifica consistenza `start_date` ↔ `end_date`**

Una volta convertite entrambe le date, verifichiamo che non ci siano righe in cui `end_date < start_date`: un anime non può terminare prima di iniziare. Consideriamo solo le righe in cui entrambe le date sono presenti, poiché confrontare con un valore nullo non ha significato.

In [22]:
# end_date < start_date
both_valid = df_details['start_date'].notna() & df_details['end_date'].notna()
inverted   = both_valid & (df_details['end_date'] < df_details['start_date'])
print(f'Righe con end_date < start_date: {inverted.sum()}')
print()
print(df_details[inverted][['mal_id', 'title', 'start_date', 'end_date']].head(10).to_string(index=False))

Righe con end_date < start_date: 101

 mal_id                                     title                start_date                  end_date
  48010                       23 Hao Niu Nai Tang 2020-03-01 00:00:00+00:00 2020-01-01 00:00:00+00:00
  23583                    Akage no Anne Specials 2001-02-11 00:00:00+00:00 2001-01-01 00:00:00+00:00
  44731                    Ali Diu Dongxi de Wawa 2012-08-31 00:00:00+00:00 2012-01-01 00:00:00+00:00
  53727                         Ashuai 8th Season 2022-06-03 00:00:00+00:00 2022-01-01 00:00:00+00:00
  45385                       AU: Kaixin Tongnian 2013-03-01 00:00:00+00:00 2013-01-01 00:00:00+00:00
  10330 Bakugan Battle Brawlers: Mechtanium Surge 2018-09-01 00:00:00+00:00 2018-01-01 00:00:00+00:00
  44551                       Baozou Kongbu Gushi 2014-07-11 00:00:00+00:00 2014-01-01 00:00:00+00:00
  44555            Baozou Kongbu Gushi 5th Season 2017-02-07 00:00:00+00:00 2017-01-01 00:00:00+00:00
  44557                                Baozo

**Osservazioni:**
- **101 righe** presentano `end_date < start_date`. Guardando i dati, il pattern è sempre lo stesso: `end_date` è il 1° gennaio di un anno (es. `2020-01-01`), mentre `start_date` è una data più tarda nello stesso anno (es. `2020-03-01`).
- La spiegazione è che a volte quando si conosce solo l'**anno** di fine trasmissione, non il giorno esatto si registra automaticamente il 1° gennaio come data di fine. Il risultato è che la data di fine appare prima della data di inizio, ma non è un errore reale, è solo una mancanza di precisione nel dato originale.
- **Nessuna pulizia necessaria.**

### 2.11 `synopsis`

Testo della sinossi dell'anime. I null sono attesi: cortometraggi, CM (Commercial Message) e PV (Promotional Video) spesso non hanno una sinossi registrata.

In [23]:
analyze(df_details['synopsis'])


  Nome serie                     synopsis
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────



  Righe totali                   28,955
  Valori non nulli               23,828  (82.29%)
  Null / NaN                     5,127  (17.71%)
  Stringhe vuote                 0
  Valori unici                   23,540  (81.30%)
  Valori duplicati               288

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  10  caratteri
  Lunghezza max                  3,750  caratteri
  Lunghezza media                377.40  caratteri
  Lunghezza mediana              251.0000  caratteri
  Dev. standard lunghezza        351.38
  IQR lunghezza                  499.0000

Valori di lunghezza estrema
────────────────────────────────────────────────────────────────────────────────

  Le 10 stringhe più lunghe:
    #1   len=3750  Li Xiaobai, who entered the immersive stand-alone game Faerie Continent, joined the Xianyu Sect of the Northern Territor…
    #2   len=3635  Tezuka × Atobe ÷ Koharu↵ Fudomine Middle Schoo

  Contiene cifre                 5,520  (23.2%)
  Contiene maiuscole             23,826  (100.0%)
  Contiene minuscole             23,828  (100.0%)
  Contiene spazi                 23,828  (100.0%)
  Contiene caratteri speciali    23,807  (99.9%)
  Contiene URL                   3  (0.0%)
  Contiene @                     38  (0.2%)
  Solo numerico                  0
  Solo alfabetico                0
  Solo alfanumerico              0

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────

  Caratteri totali               8,992,584
  Caratteri unici                617
  Caratteri medi per valore      377.40



Analisi parole / token
────────────────────────────────────────────────────────────────────────────────



  Parole medie per valore        63.48
  Max parole                     638
  Min parole                     2



  Le 10 parole più comuni:
  Parola                  Conteggio
  ────────────────────   ──────────
  the                        86,835
  to                         43,834
  of                         40,908
  a                          40,351
  and                        39,283
  in                         20,100
  is                         19,911
  by                         16,286
  his                        15,580
  for                        13,106



**Osservazioni:**
- **5.127 valori nulli** (~17.7%): strutturali — anime brevi (CM, PV, Music) che non hanno una trama e per i quali l'assenza di sinossi è attesa.
- **25 righe** contengono testo boilerplate di sistema (`"No synopsis has been added for this series yet."`): da convertire in `NaN`.
- **288 valori duplicati**: da investigare, potrebbero nascondere ulteriori pattern boilerplate non catturati dal pattern principale.
- **3 sinossi contengono URL** e **38 contengono `@`**: da investigare per verificare che si tratti di contenuto legittimo.
- Aggiungiamo una colonna `has_synopsis` dopo la pulizia del boilerplate per distinguere tra i casi con e senza sinopsi.

**Duplicati nella sinossi:**

`analyze()` riporta 288 valori duplicati. Esaminiamo i testi più ricorrenti per verificare se esistono altri pattern boilerplate non catturati dal filtro principale, oppure se si tratta di duplicati legittimi (sinossi condivise tra stagioni o titoli correlati).

In [24]:
# Duplicati in synopsis (su dati non ancora puliti)
dup_mask_syn = df_details['synopsis'].duplicated(keep=False) & df_details['synopsis'].notna()
dup_counts = df_details.loc[dup_mask_syn, 'synopsis'].value_counts()

print(f'Righe con synopsis duplicata: {dup_mask_syn.sum()}')
print(f'Testi univoci tra i duplicati: {len(dup_counts)}')
print()
print('Top 15 synopsis duplicate (prime 120 caratteri):')
print('─' * 80)
for text, cnt in dup_counts.head(15).items():
    preview = repr(text[:120].replace('\n', ' '))
    print(f'  {cnt:3}x  {preview}')

Righe con synopsis duplicata: 456
Testi univoci tra i duplicati: 168

Top 15 synopsis duplicate (prime 120 caratteri):
────────────────────────────────────────────────────────────────────────────────
   23x  'No synopsis has been added for this series yet.  Click here to update this information.'
   12x  'Furukawa Taku film.'
   12x  'Film by Takashi Ito.'
   10x  'A short animation by Taku Furukawa.'
    7x  'A short film by Okamoto Tadanari.'
    6x  'Short animation by Rapparu.'
    6x  'Short film by Hirano Ryou.'
    6x  'A short puppet animation movie by Tadahito Mochinaga.'
    6x  'Short film by Kurosaka Keita.'
    5x  'A short film by Haruka Suzuki.'
    5x  'Unaired specials included in the Blu-ray and DVD release.'
    4x  'A mix of animation and live-action.'
    4x  'Three episodes aired on TV as part of the Ribon x Oha Suta specials.'
    4x  "The music film is a part of TOHO animation's 10th anniversary project."
    4x  'Short animation by Komazaki Tomomi.'


**Risultati indagine duplicati:**

- **456 righe coinvolte** in duplicazioni (168 testi univoci distinti), più delle 288 riportate da `analyze()` perché quella cifra conta solo le copie non-prime, mentre qui si contano tutte le righe con testo non unico.
- Il testo più frequente (23×) è il boilerplate `"No synopsis has been added for this series yet..."` che gestitiamo successivamente.
- Tutti gli altri duplicati sono descrizioni brevi e generiche di cortometraggi d'autore o descrizioni standard di release. Si tratta di contenuto corretto.
- **Nessun nuovo pattern boilerplate** da rimuovere. I duplicati rimanenti sono legittimi.

**Sinossi contenenti URL e `@`:**

`analyze()` riporta 3 sinossi con URL e 38 con `@`. Verifichiamo il contesto per determinare se si tratta di contenuto legittimo o artefatti da rimuovere.

In [25]:
# Sinossi contenenti URL
url_mask_syn = df_details['synopsis'].str.contains(
    r'https?://\S+', regex=True, na=False
)
print(f'Sinossi con URL: {url_mask_syn.sum()}')
print()
cols_show = ['mal_id', 'title', 'type']
for _, row in df_details.loc[url_mask_syn, cols_show + ['synopsis']].iterrows():
    import re
    urls_found = re.findall(r'https?://\S+', row['synopsis'])
    print(f"  mal_id={row['mal_id']}  [{row['type']}]  {row['title']}")
    for u in urls_found:
        print(f"    URL: {u}")
    print()

Sinossi con URL: 3

  mal_id=47634  [Special]  Kamen Senshi Lavithunder CM
    URL: https://vgmdb.net/album/54883

  mal_id=6230  [OVA]  Kyokujitsu no Kantai
    URL: http://en.wikipedia.org/wiki/Konpeki_no_Kantai

  mal_id=43196  [Movie]  Luotuo Xian Wu
    URL: https://jouz.myblog.arts.ac.uk/,



In [26]:
# Sinossi contenenti '@'
at_mask_syn = df_details['synopsis'].str.contains(r'@', regex=False, na=False)
print(f'Sinossi con @: {at_mask_syn.sum()}')
print()

import re
pattern_at = re.compile(r'\S*@\S+')

print('Contesti con @ (titolo + frammento):')
print('─' * 80)
for _, row in df_details.loc[at_mask_syn, ['mal_id', 'title', 'synopsis']].iterrows():
    matches = pattern_at.findall(row['synopsis'])
    print(f"  mal_id={row['mal_id']}  {row['title']}")
    for m in matches:
        print(f"    @ contesto: {m}")
    print()

Sinossi con @: 38

Contesti con @ (titolo + frammento):
────────────────────────────────────────────────────────────────────────────────
  mal_id=35651  Aozora Under Girls!
    @ contesto: iDOLM@STER

  mal_id=48814  Blue
    @ contesto: m@e.

  mal_id=34489  Cinderella Girls Gekijou: 5-shuunen Kinen Short Anime
    @ contesto: iDOLM@STER

  mal_id=35346  Cinderella Girls Gekijou: Kayou Cinderella Theater
    @ contesto: iDOLM@STER

  mal_id=36500  Cinderella Girls Gekijou: Kayou Cinderella Theater 2nd Season
    @ contesto: iDOLM@STER

  mal_id=37424  Cinderella Girls Gekijou: Kayou Cinderella Theater 3rd Season
    @ contesto: iDOLM@STER

  mal_id=39654  Cinderella Girls Gekijou: Kayou Cinderella Theater 4th Season
    @ contesto: iDOLM@STER

  mal_id=51234  Eternity Memories
    @ contesto: iDOLM@STER

  mal_id=58598  Guruguru@Mawaru@Mawaluna
    @ contesto: Guruguru@Mawaru@Mawaluna

  mal_id=28439  Hatsune Miku x Tetsuya Nomura
    @ contesto: Positivity"@Wallplay

  mal_id=3406  i

**Risultati indagine URL e `@`:**

**URL (3 sinossi):**
- `mal_id=47634` (CM) — link a VGMdb: citazione della colonna sonora.
- `mal_id=6230` (OVA) — link a Wikipedia: citazione enciclopedica di approfondimento.
- `mal_id=43196` (Movie) — link a blog: citazione di fonte accademica.
- Tutti sono citazioni di fonte, non artefatti. **Nessuna pulizia necessaria.**

**`@` (38 sinossi):**
In tutti i casi `@` è un **carattere intenzionale** di titoli o nomi ufficiali. **Nessuna pulizia necessaria.**

**Sinossi boilerplate:**
Alcune righe contengono un testo boilerplate di sistema MAL invece di una vera sinossi. Questi valori vanno convertiti in `NaN`.

In [27]:
boilerplate_pattern = r'^No synopsis has been added for this series yet'
mask_bp = df_details['synopsis'].str.contains(boilerplate_pattern, regex=True, na=False)
print(f'Righe con sinossi boilerplate: {mask_bp.sum()}')
print(df_details.loc[mask_bp, 'synopsis'].value_counts().to_string())

df_details.loc[mask_bp, 'synopsis'] = np.nan
print(f'\nNull in synopsis dopo pulizia: {df_details["synopsis"].isna().sum():,}')

Righe con sinossi boilerplate: 25
synopsis
No synopsis has been added for this series yet.\n\nClick here to update this information.     23
No synopsis has been added for this series yet.                                                1
No synopsis has been added for this series yet. \n\nClick here to update this information.     1

Null in synopsis dopo pulizia: 5,152


In [28]:
df_details['has_synopsis'] = df_details['synopsis'].notna()
print(f'has_synopsis → True (con sinossi): {df_details["has_synopsis"].sum():,}  |  False: {(~df_details["has_synopsis"]).sum():,}')

has_synopsis → True (con sinossi): 23,803  |  False: 5,152


### 2.12 `rank`

Classifica globale basata sullo score. I null sono strutturali: anime senza score sufficiente non ricevono un rank. 

In [29]:
analyze(df_details['rank'])


  Nome serie                     rank
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               21,997  (75.97%)
  Null / NaN                     6,958  (24.03%)
  Zeri                           0  (0.00%)
  Positivi                       21,997   (75.97%)
  Negativi                       0   (0.00%)
  Valori unici                   17,765  (61.35%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1.0000
  Max                            22,020.00
  Range                          22,019.00
  Media                          11,033.16
  Mediana                        11,008.00
  Moda/e                         2,651.00, 2,81

**Osservazioni:**
- **6.958 valori nulli** (~24.0%): strutturali, anime non classificati (score assente o insufficiente).
- **17.765 valori unici su 21.997 non-null**: 4.232 anime condividono il rank con almeno un altro — rank *ex aequo* tipici di MAL quando più anime hanno lo stesso score. Da verificare.
- **Max rank (22.020) > non-null (21.997)**: differenza di 23 — esistono buchi nella sequenza dei rank. Possibile causa: anime rimossi da MAL nel tempo i cui rank non sono stati riassegnati. Da verificare.
- **Distribuzione quasi-uniforme**: attesa per una classifica. Leggera depressione nelle fasce 16.515–19.818 (~1.000–1.030 per bin vs ~1.100 attesi) compensata da un picco nelle ultime due fasce (1.194 e 1.218).
- **Conversione dtype**: da `float64` a `Int64` (nullable integer), dato che i rank sono interi.

**Rank ex aequo e buchi nella sequenza:**

Verifichiamo quanti anime condividono lo stesso rank e quante posizioni sono assenti nella sequenza 1–22.020.

In [30]:
non_null_rank = df_details['rank'].dropna()
n_total  = len(non_null_rank)
n_unique = non_null_rank.nunique()
n_tied   = n_total - n_unique

print(f'Rank non-null        : {n_total:,}')
print(f'Rank unici           : {n_unique:,}')
print(f'Anime in ex aequo    : {n_tied:,}  ({n_tied/n_total*100:.1f}%)')
print()

# Gruppi di ex aequo più grandi
tied_groups = non_null_rank.value_counts()
tied_groups = tied_groups[tied_groups > 1]
print(f'Gruppi ex aequo (rank condiviso): {len(tied_groups)}')
print(f'Dimensione max gruppo           : {tied_groups.max()}')
print()

# Buchi nella sequenza
rank_min = int(non_null_rank.min())
rank_max = int(non_null_rank.max())
rank_set = set(non_null_rank.astype(int))
gaps = sorted(set(range(rank_min, rank_max + 1)) - rank_set)
print(f'Range rank: {rank_min} – {rank_max}  (ampiezza: {rank_max - rank_min + 1:,})')
print(f'Buchi nella sequenza: {len(gaps)}')
if gaps:
    print(f'Posizioni mancanti: {gaps}')

Rank non-null        : 21,997
Rank unici           : 17,765
Anime in ex aequo    : 4,232  (19.2%)

Gruppi ex aequo (rank condiviso): 3956
Dimensione max gruppo           : 4

Range rank: 1 – 22020  (ampiezza: 22,020)
Buchi nella sequenza: 4255
Posizioni mancanti: [121, 130, 154, 170, 177, 183, 185, 240, 254, 285, 288, 291, 293, 349, 410, 422, 424, 428, 444, 446, 448, 451, 469, 476, 571, 575, 576, 580, 586, 591, 603, 610, 615, 631, 635, 639, 642, 644, 688, 691, 694, 701, 709, 710, 714, 719, 722, 724, 728, 740, 743, 758, 791, 850, 923, 931, 933, 939, 1000, 1019, 1026, 1029, 1032, 1042, 1046, 1048, 1051, 1058, 1061, 1065, 1156, 1158, 1162, 1166, 1214, 1217, 1229, 1235, 1240, 1243, 1247, 1248, 1253, 1274, 1289, 1290, 1292, 1294, 1302, 1305, 1330, 1334, 1345, 1347, 1360, 1365, 1377, 1379, 1380, 1384, 1397, 1407, 1432, 1440, 1443, 1445, 1449, 1453, 1457, 1461, 1464, 1471, 1477, 1486, 1489, 1496, 1500, 1505, 1507, 1509, 1512, 1516, 1519, 1520, 1524, 1527, 1533, 1537, 1543, 1547, 1563, 1582, 1

**Risultati indagine rank ex aequo e buchi:**

- **4.232 anime in ex aequo** (19.2% dei non-null): 3.956 gruppi di rank condiviso, con dimensione massima di 4. Comportamento atteso: MAL assegna lo stesso rank a tutti gli anime con score identico.
- **4.255 buchi nella sequenza** (range 1–22.020): conseguenza diretta del meccanismo di *standard competition ranking* usato da MAL. Quando k anime sono a pari merito al rank r, le k−1 posizioni successive (r+1 … r+k−1) vengono saltate. In formula: buchi ≈ n_anime_in_ex_aequo + (max_rank − n_non_null) = 4.232 + 23 = 4.255.
- I 23 gap "eccedenti" (max_rank 22.020 − non-null 21.997) suggeriscono che alcuni anime potrebbero essere stati rimossi da MAL nel tempo senza riassegnazione dei rank.
- **Nessuna pulizia necessaria**: i buchi e gli ex aequo sono strutturali e non indicano errori nei dati.

In [31]:
df_details['rank'] = df_details['rank'].astype('Int64')
print(f'rank dtype   : {df_details["rank"].dtype}')
print(f'Null residui : {df_details["rank"].isna().sum():,}')

rank dtype   : Int64
Null residui : 6,958


### 2.13 `popularity`, `members`, `favorites`

Tre colonne numeriche che misurano l'interesse degli utenti MAL verso l'anime. Sono sempre valorizzate (anche gli anime senza voti hanno qualche utente nella lista). Il dtype `int64` è già corretto.

In [32]:
analyze(df_details['popularity'])
analyze(df_details['members'])
analyze(df_details['favorites'])


  Nome serie                     popularity
  dtype                          int64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Zeri                           0  (0.00%)
  Positivi                       28,955   (100.00%)
  Negativi                       0   (0.00%)
  Valori unici                   21,954  (75.82%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1
  Max                            28,999
  Range                          28,998
  Media                          14,500.36
  Mediana                        14,497.00
  Moda/e                         10,785, 15,862, 21,379
 

  Soglia bassa  (Q1 − 1.5×IQR)   -13,207.75
  Soglia alta   (Q3 + 1.5×IQR)   22,634.25
  Outlier rilevati               5,042  (17.41%)

  Top 5 outlier estremi (per valore assoluto):
    [22010]  4,230,312
    [4788]  4,170,767
    [6593]  3,576,199
    [17990]  3,430,848
    [12300]  3,323,322

Distribuzione
────────────────────────────────────────────────────────────────────────────────

  Intervallo                      Conteggio        %
  ────────────────────────────── ──────────   ──────
  23 – 211,537                       27,690    95.6%
  211,537 – 423,051                     663     2.3%
  423,051 – 634,566                     227     0.8%
  634,566 – 846,080                     131     0.5%
  846,080 – 1,057,595                    78     0.3%
  1,057,595 – 1,269,109                  48     0.2%
  1,269,109 – 1,480,624                  24     0.1%
  1,480,624 – 1,692,138                  23     0.1%
  1,692,138 – 1,903,653                  25     0.1%
  1,903,653 – 2,115,167

  Min                            0
  Max                            243,358
  Range                          243,358
  Media                          430.8490
  Mediana                        1.0000
  Moda/e                         0
  Dev. standard                  4,520.61
  Varianza                       20,435,916
  Errore std (SEM)               26.5666
  IQR  (Q3 − Q1)                 17.0000
  Q1  (25° pct)                  0.0000
  Q3  (75° pct)                  17.0000
  P5  ( 5° pct)                  0.0000
  P95 (95° pct)                  806.9000
  Coefficiente di variazione     1049.23%

Analisi outlier
────────────────────────────────────────────────────────────────────────────────

  Soglia bassa  (Q1 − 1.5×IQR)   -25.5000
  Soglia alta   (Q3 + 1.5×IQR)   42.5000
  Outlier rilevati               5,336  (18.43%)

  Top 5 outlier estremi (per valore assoluto):
    [17919]  243,358
    [6593]  236,754
    [9671]  223,927
    [23288]  198,245
    [22010]  183,305

Distribuzi

**Osservazioni:**
- **Nessun valore nullo** in nessuna delle tre colonne.
- **Dtype corretto**: `int64` per tutte e tre.
- **Distribuzioni asimmetriche**: code lunghe verso destra, da tenere in considerazione nelle analisi.

Nessuna pulizia necessaria.

In [33]:
# Nessuna operazione richiesta
print(df_details[['popularity', 'members', 'favorites']].isnull().sum().to_string())

popularity    0
members       0
favorites     0


### 2.14 `genres`, `studios`, `themes`, `demographics`, `producers`, `licensors`, `streaming`

Colonne contenenti liste serializzate come stringhe Python (es. `"['Comedy', 'Drama']"`).

In [34]:
list_cols = ['genres', 'studios', 'themes', 'demographics', 'producers', 'licensors', 'streaming']

summary = pd.DataFrame({
    'null'    : df_details[list_cols].isnull().sum(),
    'unici'   : df_details[list_cols].nunique(),
    'solo []' : (df_details[list_cols] == '[]').sum(),
}).rename_axis('colonna')
print(summary.to_string())

              null  unici  solo []
colonna                           
genres           0    931     5983
studios          0   1910    11703
themes           0   1044    11818
demographics     0      9    18140
producers        0   5306    15272
licensors        0    304    23787
streaming        0    434    25197


**Osservazioni:**
- **Nessun valore nullo** in nessuna delle 7 colonne.
- **Liste vuote `[]` legittime**: anime senza studio, senza genere, senza streaming, ecc.
- **Formato stringa mantenuto**: sufficiente per le analisi testuali; deserializzabile con `ast.literal_eval` quando necessario.

Nessuna pulizia necessaria.

In [35]:
# Nessuna operazione richiesta
list_cols = ['genres', 'studios', 'themes', 'demographics', 'producers', 'licensors', 'streaming']
print(df_details[list_cols].isnull().sum().to_string())

genres          0
studios         0
themes          0
demographics    0
producers       0
licensors       0
streaming       0


### 2.15 `source`

Colonna categorica con il materiale di origine dell'anime (Manga, Novel, Original, …). I valori sono coerenti con la tassonomia MAL; non ci aspettiamo null.

In [36]:
analyze(df_details['source'])


  Nome serie                     source
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   17  (0.06%)
  Valori duplicati               28,938

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  4  caratteri
  Lunghezza max                  12  caratteri
  Lunghezza media                7.28  caratteri
  Lunghezza mediana              8.0000  caratteri
  Dev. standard lunghezza        2.11
  IQR lunghezza                  3.0000

Valori di lunghezza estrema
───────────────────────

  Contiene cifre                 339  (1.2%)
  Contiene maiuscole             28,616  (98.8%)
  Contiene minuscole             28,955  (100.0%)
  Contiene spazi                 4,174  (14.4%)
  Contiene caratteri speciali    339  (1.2%)
  Contiene URL                   0  (0.0%)
  Contiene @                     0  (0.0%)
  Solo numerico                  0
  Solo alfabetico                24,781
  Solo alfanumerico              24,781

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────

  Caratteri totali               210,660
  Caratteri unici                35
  Caratteri medi per valore      7.28


**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un valore di source.
- **17 valori distinti**: tutti coerenti con la tassonomia MAL, incluso `Unknown` (2.794 casi) che è un valore esplicito per anime con origine non documentata.

Nessuna pulizia necessaria.

In [37]:
# Nessuna operazione richiesta
print(f'Null in source : {df_details["source"].isna().sum()}')
print(df_details['source'].value_counts())

Null in source : 0
source
Original        12359
Manga            5508
Unknown          2794
Game             1473
Other            1326
Light novel      1199
Visual novel     1160
Novel             820
Web manga         654
4-koma manga      339
Picture book      284
Music             263
Mixed media       237
Book              224
Web novel         222
Card game          79
Radio              14
Name: count, dtype: int64


### 2.16 `rating`

Classificazione per età secondo la tassonomia MAL (G, PG, PG-13, R-17+, R+, Rx). I null sono strutturali: alcuni anime non hanno ancora una classificazione registrata.

In [38]:
analyze(df_details['rating'])


  Nome serie                     rating
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,355  (97.93%)
  Null / NaN                     600  (2.07%)
  Stringhe vuote                 0
  Valori unici                   6  (0.02%)
  Valori duplicati               28,349



Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  11  caratteri
  Lunghezza max                  30  caratteri
  Lunghezza media                18.08  caratteri
  Lunghezza mediana              13.0000  caratteri
  Dev. standard lunghezza        6.67
  IQR lunghezza                  13.0000

Valori di lunghezza estrema
────────────────────────────────────────────────────────────────────────────────

  Le 10 stringhe più lunghe:
    #1   len=30  R - 17+ (violence & profanity)
    #2   len=30  R - 17+ (violence & profanity)
    #3   len=30  R - 17+ (violence & profanity)
    #4   len=30  R - 17+ (violence & profanity)
    #5   len=30  R - 17+ (violence & profanity)
    #6   len=30  R - 17+ (violence & profanity)
    #7   len=30  R - 17+ (violence & profanity)
    #8   len=30  R - 17+ (violence & profanity)
    #9   len=30  R - 17+ (violence & profanity)
    #10  len=30  R - 17+ (violence & profan

  Caratteri totali               512,707
  Caratteri unici                37
  Caratteri medi per valore      18.08


Analisi parole / token
────────────────────────────────────────────────────────────────────────────────



  Parole medie per valore        4.64
  Max parole                     6
  Min parole                     3

  Le 10 parole più comuni:
  Parola                  Conteggio
  ────────────────────   ──────────
  -                          28,355
  PG-13                      10,463
  Teens                      10,463
  13                         10,463
  or                         10,463
  older                      10,463
  G                           9,066
  All                         9,066
  Ages                        9,066
  PG                          4,421



**Osservazioni:**
- **600 valori nulli** (~2.1%): strutturali, anime senza classificazione per età registrata su MAL.
- **6 valori distinti**: G, PG, PG-13, R-17+, R+, Rx, tutti coerenti con la classificazione MAL.

Nessuna pulizia necessaria.

In [39]:
# Nessuna operazione richiesta: i null sono strutturali
print(f'Null in rating : {df_details["rating"].isna().sum()}')
print(df_details['rating'].value_counts())

Null in rating : 600
rating
PG-13 - Teens 13 or older         10463
G - All Ages                       9066
PG - Children                      4421
R - 17+ (violence & profanity)     1593
Rx - Hentai                        1583
R+ - Mild Nudity                   1229
Name: count, dtype: int64


### 2.17 `episodes`

Numero di episodi dell'anime. I null sono strutturali: anime in corso o non ancora in onda possono non avere ancora un numero definitivo. Il dtype `float64` va convertito a `Int64` (nullable integer).

In [40]:
analyze(df_details['episodes'])


  Nome serie                     episodes
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,275  (97.65%)
  Null / NaN                     680  (2.35%)
  Zeri                           0  (0.00%)
  Positivi                       28,275   (97.65%)
  Negativi                       0   (0.00%)
  Valori unici                   261  (0.90%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1.0000
  Max                            3,000.00
  Range                          2,999.00
  Media                          14.1301
  Mediana                        2.0000
  Moda/e                         1.0000
  Dev. standard  

**Osservazioni:**
- **680 valori nulli** (~2.3%): strutturali, anime in corso o non ancora in onda senza numero definitivo di episodi.
- **Conversione dtype**: da `float64` a `Int64` (nullable integer), dato che i valori sono interi.

In [41]:
df_details['episodes'] = df_details['episodes'].astype('Int64')
print(f'episodes dtype   : {df_details["episodes"].dtype}')
print(f'Null residui     : {df_details["episodes"].isna().sum()}')
print(f'Range            : {df_details["episodes"].min()} – {df_details["episodes"].max()}')

episodes dtype   : Int64
Null residui     : 680
Range            : 1 – 3000


### 2.18 `season`

Stagione di uscita dell'anime (spring, summer, fall, winter). I null sono strutturali: Movies, OVA, ONA, Music, CM e PV non appartengono a una stagione MAL.

In [42]:
analyze(df_details['season'])


  Nome serie                     season
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               6,266  (21.64%)
  Null / NaN                     22,689  (78.36%)
  Stringhe vuote                 0
  Valori unici                   4  (0.01%)
  Valori duplicati               6,262

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  4  caratteri
  Lunghezza max                  6  caratteri
  Lunghezza media                5.42  caratteri
  Lunghezza mediana              6.0000  caratteri
  Dev. standard lunghezza        0.91
  IQR lunghezza                  2.0000

Valori di lunghezza estrema
──────────────────────

**Osservazioni:**
- **22.689 valori nulli** (~78.4%): strutturali, Movies, OVA, ONA, Music, CM e PV non appartengono a una stagione MAL.
- **4 valori distinti**: `spring`, `summer`, `fall`, `winter`, tutti coerenti.

Nessuna pulizia necessaria.

In [43]:
# Nessuna operazione richiesta: i null sono strutturali
print(f'Null in season : {df_details["season"].isna().sum():,} ({df_details["season"].isna().mean()*100:.1f}%)')
print(df_details['season'].value_counts())

Null in season : 22,689 (78.4%)
season
spring    1972
fall      1825
winter    1302
summer    1167
Name: count, dtype: int64


### 2.19 `year`

Anno di uscita dell'anime. I null corrispondono esattamente a quelli di `season` (stessa logica strutturale). Il dtype `float64` va convertito a `Int64` (nullable integer).

In [44]:
analyze(df_details['year'])


  Nome serie                     year
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               6,266  (21.64%)
  Null / NaN                     22,689  (78.36%)
  Zeri                           0  (0.00%)
  Positivi                       6,266   (21.64%)
  Negativi                       0   (0.00%)
  Valori unici                   66  (0.23%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1,961.00
  Max                            2,026.00
  Range                          65.0000
  Media                          2,010.03
  Mediana                        2,014.00
  Moda/e                         2,023.00
  Dev. standar

**Osservazioni:**
- **22.689 valori nulli**: identici a quelli di `season` (stessa logica strutturale).
- **Conversione dtype**: da `float64` a `Int64` (nullable integer).

In [45]:
df_details['year'] = df_details['year'].astype('Int64')
print(f'year dtype   : {df_details["year"].dtype}')
print(f'Null residui : {df_details["year"].isna().sum():,}')
print(f'Range        : {df_details["year"].min()} – {df_details["year"].max()}')

year dtype   : Int64
Null residui : 22,689
Range        : 1961 – 2026


In [46]:
year_from_date = df_details['start_date'].dt.year
mismatch = df_details['year'].notna() & df_details['start_date'].notna() & \
           (df_details['year'].astype('Int64') != year_from_date.astype('Int64'))
print(f'year != start_date.year: {mismatch.sum()}')
print()
print(df_details[mismatch][['mal_id', 'title', 'year', 'season', 'start_date']].head(10).to_string(index=False))

year != start_date.year: 84

 mal_id                                   title  year season                start_date
  53381             Agisang-eo Ollie wa William  2021 winter 2020-12-25 00:00:00+00:00
  46296                              Ao Ao Long  2011 winter 2010-12-01 00:00:00+00:00
  61333      Ao no Miburo: Serizawa Ansatsu-hen  2026 winter 2025-12-20 00:00:00+00:00
   1550                             Attack No.1  1970 winter 1969-12-07 00:00:00+00:00
   3166                                Ayakashi  2008 winter 2007-12-12 00:00:00+00:00
  57960 Balala Xiao Mo Xian: Xing Yuan Die Qi 2  2024 winter 2023-12-28 00:00:00+00:00
  48212                           Bangle School  2018 winter 2017-12-02 00:00:00+00:00
   1667                               Barom One  2003 winter 2002-12-07 00:00:00+00:00
  53372                       Beuledeu Ibalso 3  2022 winter 2021-12-25 00:00:00+00:00
  16405          Boku no Imouto wa "Osaka Okan"  2013 winter 2012-12-22 00:00:00+00:00


**Osservazioni:**
- **84 righe** in cui `year` ≠ anno di `start_date`. Tutti i casi riguardano anime `winter`: la stagione invernale MAL abbraccia dicembre–marzo, quindi un anime classificato come "winter 2021" può avere `start_date` a dicembre 2020. La discrepanza è attesa e coerente con la convenzione MAL. **Nessuna pulizia necessaria.**

### 2.20 `explicit_genres`

Colonna destinata ai generi espliciti (es. Hentai). Verifichiamo se contiene informazione utile o se è interamente vuota.

In [47]:
analyze(df_details['explicit_genres'])


  Nome serie                     explicit_genres
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Stringhe vuote                 0
  Valori unici                   1  (0.00%)
  Valori duplicati               28,954

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  2  caratteri
  Lunghezza max                  2  caratteri
  Lunghezza media                2.00  caratteri
  Lunghezza mediana              2.0000  caratteri
  Dev. standard lunghezza        0.00
  IQR lunghezza                  0.0000

Valori di lunghezza estrema
────────────────

  Contiene cifre                 0  (0.0%)
  Contiene maiuscole             0  (0.0%)
  Contiene minuscole             0  (0.0%)
  Contiene spazi                 0  (0.0%)
  Contiene caratteri speciali    28,955  (100.0%)
  Contiene URL                   0  (0.0%)
  Contiene @                     0  (0.0%)
  Solo numerico                  0
  Solo alfabetico                0
  Solo alfanumerico              0

Statistiche a livello di carattere
────────────────────────────────────────────────────────────────────────────────

  Caratteri totali               57,910
  Caratteri unici                2
  Caratteri medi per valore      2.00


**Osservazioni:**
- La colonna contiene esclusivamente il valore `[]` (lista vuota) per tutte le 28.955 righe: **nessuna informazione utile**.
- **Rimozione della colonna**.

In [48]:
print(f'Valori unici in explicit_genres: {df_details["explicit_genres"].unique()}')
df_details.drop(columns=['explicit_genres'], inplace=True)
print('Colonna explicit_genres rimossa.')

Valori unici in explicit_genres: <StringArray>
['[]']
Length: 1, dtype: str
Colonna explicit_genres rimossa.


### 2.21 Chiave primaria `mal_id`

Verifichiamo che `mal_id` sia ancora univoco dopo tutte le operazioni di pulizia.

In [49]:
n_pk_dup = df_details.duplicated(subset=['mal_id'], keep=False).sum()
print(f'Duplicati su mal_id: {n_pk_dup}')

if n_pk_dup > 0:
    print('→ Rimozione duplicati sulla chiave primaria...')
    df_details.drop_duplicates(subset=['mal_id'], keep='first', inplace=True)
    print(f'Righe dopo la rimozione: {len(df_details):,}')
else:
    print('→ Chiave primaria già univoca, nessuna operazione richiesta.')

Duplicati su mal_id: 0
→ Chiave primaria già univoca, nessuna operazione richiesta.


## 3. Riepilogo e Salvataggio

Le operazioni di pulizia sono state effettuate colonna per colonna nella sezione 2. In questa sezione riepiloghiamo il risultato ed effettuiamo il salvataggio del dataset finale.

In [50]:
print('=== Riepilogo Dataset Pulito ===')
print(f'Righe originali      : {n_originale:>10,}')
print(f'Righe dopo cleaning  : {len(df_details):>10,}')
print(f'Righe rimosse totali : {n_originale - len(df_details):>10,}')
print()
print('Operazioni di pulizia effettuate:')
print(f'  - explicit_genres rimossa (colonna vuota)')
print(f'  - scored_by, rank, episodes, year → Int64')
print(f'  - start_date, end_date → datetime64[us, UTC]')
print(f'  - 25 sinossi boilerplate → NaN')
print(f'  - has_image, has_synopsis aggiunte')
print()
print('Dtypes finali:')
print(df_details.dtypes)
print()
print('Valori mancanti residui per colonna:')
null_summary = df_details.isnull().sum()
print(null_summary[null_summary > 0].to_string())

=== Riepilogo Dataset Pulito ===
Righe originali      :     28,955
Righe dopo cleaning  :     28,955
Righe rimosse totali :          0

Operazioni di pulizia effettuate:
  - explicit_genres rimossa (colonna vuota)
  - scored_by, rank, episodes, year → Int64
  - start_date, end_date → datetime64[us, UTC]
  - 25 sinossi boilerplate → NaN
  - has_image, has_synopsis aggiunte

Dtypes finali:
mal_id                          int64
title                             str
title_japanese                    str
url                               str
image_url                         str
type                              str
status                            str
score                         float64
scored_by                       Int64
start_date        datetime64[us, UTC]
end_date          datetime64[us, UTC]
synopsis                          str
rank                            Int64
popularity                      int64
members                         int64
favorites                       int64
g

title_japanese      119
type                 67
score             10073
scored_by         10073
start_date          851
end_date          17688
synopsis           5152
rank               6958
rating              600
episodes            680
season            22689
year              22689


In [51]:
df_details.head()

,mal_id,title,title_japanese,url,image_url,type,status,score,scored_by,start_date,...,source,rating,episodes,season,year,producers,licensors,streaming,has_image,has_synopsis
0,59356,-Socket-,-socket-,https://myanimelist.net/anime/59356/-Socket-,https://cdn.myanimelist.net/images/anime/1043/...,Movie,Finished Airing,NaN,<NA>,2010-01-01 00:00:00+00:00,...,Original,G - All Ages,1,NaN,<NA>,['Nagoya Zokei University'],[],[],True,True
1,56036,......,......,https://myanimelist.net/anime/56036/-,https://cdn.myanimelist.net/images/anime/1057/...,Music,Finished Airing,6.53,503,2023-06-11 00:00:00+00:00,...,Original,PG-13 - Teens 13 or older,1,NaN,<NA>,[],[],[],True,True
2,2928,.hack//G.U. Returner,.HACK//G.U. RETURNER,https://myanimelist.net/anime/2928/hack__GU_Re...,https://cdn.myanimelist.net/images/anime/1798/...,OVA,Finished Airing,6.65,9745,2007-01-18 00:00:00+00:00,...,Game,PG-13 - Teens 13 or older,1,NaN,<NA>,"['Bandai Visual', 'CyberConnect2']",[],[],True,True
3,3269,.hack//G.U. Trilogy,.hack//G.U. Trilogy,https://myanimelist.net/anime/3269/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/1566/...,Movie,Finished Airing,7.06,15373,2007-12-22 00:00:00+00:00,...,Game,PG-13 - Teens 13 or older,1,NaN,<NA>,['Bandai Visual'],"['Funimation', 'Bandai Entertainment']",[],True,True
4,4469,.hack//G.U. Trilogy: Parody Mode,.hack//G.U. Trilogy,https://myanimelist.net/anime/4469/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/10/86...,Special,Finished Airing,6.35,4317,2008-03-25 00:00:00+00:00,...,Game,PG-13 - Teens 13 or older,1,NaN,<NA>,['Bandai Visual'],[],[],True,True


In [52]:
df_details.to_csv('../datasets_cleaned/details_clean.csv', index=False)
print('Salvato: datasets_cleaned/details_clean.csv')

Salvato: datasets_cleaned/details_clean.csv
